## Tugas Praktikum Individu Pemrograman Berbasis Fungsi
Nama : Rafi Fadhlillah <br>
NIM : 121450143 <br>
Kelas : RC

# 2 Pemrograman python berbasis fungsi

In [1]:
def square (x):
  y=x*x
  return y

## 2.1 Jenis fungsi dalam python


In [2]:
#built in fungction max() -> mengeluarkan nilai paling tinggi dalam suatu list
L=[1,2,3,8,5,98,1,0,-2]
max(L)

98

In [3]:
def cube(x):
    return x*x*x
cube(10)

1000

In [4]:
import math #modul yang di import adalah math
math.sqrt(100)  #fungsi sqrt dalam modul math


10.0

## 2.2 Fungsi sebagai cara modularisasi algoritma


In [5]:
#code python untuk menghitung faktor dari sebuah bilagan n
n=64
fakctors=[]
for i in range(1,n+1):
    if n % i == 0:
        fakctors.append(i)
print(fakctors)

[1, 2, 4, 8, 16, 32, 64]


In [6]:
#modularisasi code nya
def habis_dibagi(n,i):
    return n % i == 0
def faktor(n):
    temp =[]
    for i in range(1,n+1):
        if habis_dibagi(n,i):
            temp.append(i)
    return temp
faktor(64)

[1, 2, 4, 8, 16, 32, 64]

In [7]:
faktor(25)

[1, 5, 25]

## 2.3 Anonymous fungction

In [8]:
#normal user defined function
def rata(L):
    res=0
    for el in L:
        res =res +el
    res=res/len(L)
    return res 
L=[1,2,3,4,5,6,7,8,9,10]
rata(L)

5.5

In [9]:
#lambda fungcton
rata_lambda =lambda L : sum(L)/len(L)
rata_lambda(L)

5.5

In [10]:
greet = lambda:'Hello World'
greet ()

'Hello World'

In [11]:
greet= lambda name,gender :'hello '+('Mr.' if gender=='M' else 'Mrs.')+ name 
greet('Alvin','M')

'hello Mr.Alvin'

In [12]:
def helper(L):
    res=0
    for el in L:
        res=res +el
    return res 
rata=lambda L:helper(L)/len(L)
rata(L)

5.5

## 2.4 Ternary expression and operator

In [13]:
#mencari minimum dari 2 bilangan
a,b=10,20
if a>b: 
    min =b
else: 
    min =a
min

10

In [14]:
#mencari minimum dari 2 bilangan mengguakan operator ternary
a,b = 10,20
minim = a if a < b else b
print(minim)

10


In [15]:
#menggunakan nested if-else
a,b = 10,20
if a != b:
    if a>b:
        print("a lebih besar dari b")
    else :
        print("b lebih besar dari a")
else :
    print("a dan b bernilai saman")

b lebih besar dari a


In [16]:
#nested ternary operator 
a,b =10,20
print("a dan b bernilai sama" if a == b else " a lebih besar dari b "
if a > b else "b lebih besar dari a")

b lebih besar dari a


> ternary operator

In [17]:
#mengecek bilagan genap atau ganjil
f = lambda x:'genap' if x%2==0 else 'ganjil'
f(10)

'genap'

## 2.5 lambda fungction direct calling

In [18]:
(lambda x: 'genap' if x%2==0 else 'ganjil')(10)

'genap'

In [19]:
val =[1,2,3,4,5,6]
from functools import reduce

#multiply every item by two
list(map(lambda x: x * 2, val)) #[2,4,6,8,10,12]
#take the faktorial by multiplying the value so far the next item
reduce(lambda  x,y: x * y,val, 1) #1*1*2*3*4*5*6

720

# 3. Pemrograman berbasis fungsi pada analisis data

## 3.1.1 Data preprocessing dengan pemrograman prosedural

In [20]:
import pandas as pd
import numpy as np
np.random.seed(42)

def create_random_dataframe(nb_rows=1000,percent_na_per_column=[0.01,0.1,0.-9]):
    df=pd.DataFrame()
    for i,percent_na in enumerate(percent_na_per_column):
        values = [np.random.random()if np.random.random() > percent_na else np.nan for _ in range(nb_rows)]
        df[f'val_{i}']=values
        df.index=pd.date_range('2023/02/14',periods=nb_rows,freq='H')
        return df

def preprocess(df,na_percentage = 0.2, resampling_str='2H'):
    #remove columns with too many nas
    na_df = df.isna().sum()/len(df)
    list_col_to_keep = na_df[na_df <na_percentage].index
    df = df [list_col_to_keep]

    #resampling 
    df =df.resample(resampling_str).mean()

    #filling missing values
    df = df.interpolate().ffill().bfill()
    return df

random_df = create_random_dataframe()
res = preprocess(random_df)
res 

val_0
2023-02-14 00:00:00  0.774686
2023-02-14 02:00:00  0.511085
2023-02-14 04:00:00  0.838991
2023-02-14 06:00:00  0.197872
2023-02-14 08:00:00  0.407993
...                       ...
2023-03-27 06:00:00  0.674888
2023-03-27 08:00:00  0.685168
2023-03-27 10:00:00  0.822288
2023-03-27 12:00:00  0.281785
2023-03-27 14:00:00  0.612985

[500 rows x 1 columns]

In [21]:
import time
start_time = time.time()
create_random_dataframe(1000000)
end_time = time.time()
prosedural_time_taken = end_time - start_time
print("Pemrograman Prosedural: ", prosedural_time_taken, "seconds")

Pemrograman Prosedural:  0.9239468574523926 seconds


## 3.1.2 data preprocessing dengan pemrograman fungsi

In [22]:
import pandas as pd
import numpy as np
np.random.seed(42)

def create_random_dataframe(nb_rows=1000,percent_na_per_column=[0.01,0.1,0.-9]):
    df=pd.DataFrame()
    for i,  percent_na in enumerate(percent_na_per_column):
        values = [np.random.random()if np.random.random() > percent_na else np.nan for _ in range(nb_rows)]
        df[f'val_{i}']=values
        df.index=pd.date_range('2023/02/14',periods=nb_rows,freq='H')
        return df

def create_col_with_too_many_nas_remover(na_percentage =0.2):
    def remover(df):
        na_df = df.isna().sum()/len(df)
        list_col_to_keep = na_df[na_df <na_percentage].index
        return df [list_col_to_keep]
    return remover 

def create_resampler(resampling_str):
    def resampler(df):
        return df.resample(resampling_str).mean()
    return resampler

def fill_na(df):
    return df.interpolate().ffill().bfill()

random_df = create_random_dataframe()
remover =create_col_with_too_many_nas_remover()
resampler = create_resampler('2H')
res = fill_na(resampler(remover(random_df)))
res

val_0
2023-02-14 00:00:00  0.774686
2023-02-14 02:00:00  0.511085
2023-02-14 04:00:00  0.838991
2023-02-14 06:00:00  0.197872
2023-02-14 08:00:00  0.407993
...                       ...
2023-03-27 06:00:00  0.674888
2023-03-27 08:00:00  0.685168
2023-03-27 10:00:00  0.822288
2023-03-27 12:00:00  0.281785
2023-03-27 14:00:00  0.612985

[500 rows x 1 columns]

In [23]:
import time
start_time = time.time()
create_random_dataframe(1000000)
end_time = time.time()
fungsional_time_taken = end_time - start_time
print("Pemrograman fungsi : ", fungsional_time_taken, "seconds")

Pemrograman fungsi :  0.9051780700683594 seconds


## Analisis code  sebelum dan sesudah di transformasi  menjadi fungctional  programming 

> Setelah transformasi ada nya perbedaan dari fungsi 

## Analisis

Pada Pemrograman Prodsedural terdapat 2 fungsi sedangkan pada Pemrograman Fungsional terdapat 6 fungsi. Pemrograman fungsional mengacu pada kemampuan untuk memperlakukan fungsi sebagai nilai.Pemrograman Prosedural mengacu pada kemampuan untuk merangkum urutan umum instruksi ke dalam prosedur sehingga instruksi tersebut dapat dipanggil dari banyak tempat.  Pemrograman fungsional lebih modular. Dalam kode tersebut, mendefinisikan tiga fungsi terpisah yang merangkum tiga langkah preprocessing: create_col_with_too_many_nas_remover, create_resampler, dan fill_na.Pada pemrograman prosedural mendefinisikan preprocess fungsi tunggal yang melakukan semua langkah preprocessing dalam fungsi yang sama. Meskipun lebih ringkas, ini bisa menjadi kurang fleksibel jika langkah pra pemrosesan tambahan perlu ditambahkan atau dihapus di masa mendatang.Hal tersebut dapat dibuktikan dengan melihat hasil test runtime code, didapatkan waktu sebesar sebagai berikut :<br> 
Pemrograman Prosedural : 0.9239468574523926 seconds <br>
Pemrograman Fungsional : 0.9051780700683594 seconds <br>

Sehingga dapat disimpulkan bahwa pemrograman fungsional memiliki runtime yang lebih cepat daripada pemrograman prosedural. 

## 3.1.3 Bagaimana mengubah prosedural programing ke fungctional programing

In [24]:
#prosedural code
starting_number =96

#get the square of the number
square =starting_number ** 2

#increment the number by 1
increment =square + 1

#cube of the number
cube = increment ** 3

#decrase the cube by 1
decrement =cube - 1

#get the final result
result =print(decrement)

783012621312


In [25]:
#define a fungction 'call' where provide the  fungction and the arguments
def call(x,f):
  return f(x)

#define a fungction that returns the square
square = lambda x: x*x 

#define a  fungction that returns the increment
increment =lambda x:x+1

#define a fungction that returns the cube
cube = lambda x: x*x*x

#define a fungction that returns the decrement
decrement = lambda x: x-1

#put all fungction.below in the non fungctional part
funcs =[square,increment,cube,decrement]

#bring it all together.bellow is the non fungctional part
#in fungctional programming you separate you the fungction and the non fungction part
from functools import reduce  #recude is in the functools library
print(reduce(call,funcs,96))

783012621312
